src/Adapter/Product/Stock/Repository/StockMovementRepository.php

```diff
@@ -56,25 +54,113 @@ public function __construct(
+                'MIN(employee_firstname) employee_firstname',
+                'MIN(employee_lastname) employee_lastname',
+                'SUM(sm.sign * sm.physical_quantity) delta_quantity',
+                'MIN(sm.date_add) date_add_min',
+                'MAX(sm.date_add) date_add_max'
+            )
+            ->orderBy('id_stock_mvt_min', 'DESC')
+        ;
+        $this->updateQueryBuilderWithGroupings(
+            $queryBuilder,
+            $historySettings->getSingleFilter()
+        );
+        if ($historySettings->isZeroQuantityGroupingExcluded()) {
+            $queryBuilder->andHaving('delta_quantity != 0');
+        }
+
+        return $queryBuilder->execute()->fetchAllAssociative();
+    }
+
+    /**
+     * Returns a new query builder with the given filter.
+     */
+    protected function createFilterQueryBuilder(
+        StockMovementFilter $filter,
+        int $offset = 0,
+        int $limit = self::DEFAULT_LIMIT
+    ): QueryBuilder {
+        $queryBuilder = $this
+            ->connection
+            ->createQueryBuilder()
+            ->select('sm.*')
             ->from($this->dbPrefix . 'stock_mvt', 'sm')
-            ->where('id_stock = :stockId')
-            ->setParameter('stockId', $stockId->getValue())
-            ->addOrderBy('sm.date_add', 'DESC')
-            ->addOrderBy('sm.id_stock_mvt', 'DESC')
             ->setFirstResult($offset)
             ->setMaxResults($limit)
         ;
+        if (!empty($filter->getStockIds())) {
+            $queryBuilder
+                ->andWhere('sm.id_stock IN (:stockIds)')
+                ->setParameter('stockIds', $filter->getStockIdsAsString())
+            ;
+        }
+        if (null !== $filter->getIsOrder()) {
+            $queryBuilder->andWhere(
+                'sm.id_order IS ' . ($filter->getIsOrder() ? 'NOT NULL' : 'NULL')
+            );
+        }
+
+        return $queryBuilder;
+    }
 
-        return $qb->execute()->fetchAllAssociative();
+    /**
+     * Updates a query builder with a filter to group rows together
+     *
+     * @param QueryBuilder $queryBuilder Query builder to update
+     * @param StockMovementFilter $singleFilter Filter to exclude single rows from groupings
+     */
+    protected function updateQueryBuilderWithGroupings(
+        QueryBuilder $queryBuilder,
+        StockMovementFilter $singleFilter
+    ): void {
+        $pkColumn = 'sm.id_stock_mvt';
+        $groupingIdColumn = 'grouping_id';
+        $groupingNameColumn = 'grouping_name';
+        $groupingQueryBuilder = $this->createFilterQueryBuilder($singleFilter);
+        $groupingCondition = (string) $groupingQueryBuilder->getQueryPart('where');
+        $queryBuilder
+            ->addSelect(
+                implode(' ', [
+                    "MIN(@$groupingIdColumn := CASE",
+                    "WHEN @$groupingIdColumn IS NULL THEN $pkColumn",
+                    "WHEN $groupingCondition THEN $pkColumn",
```

Comment:
Variable interpoling in strings is a native feature in PHP. Nothing mysterious.
"We never did that" is pretty much the same as: "Don't change anything because we always did another way"
I assume the "mind" reason is more a personal preference than a real improvement here.
 
I already answered your previous comment, so I repeat. Since you are talking about making the code easier to read and to understand, `sprintf(`) syntax is not the best to understand a complex SQL query like this one.
Just give it a try: replace variables and take a first aspirin. Now take a second aspirin, and try to understand the `printf` syntax...
Since you insist I will change the code, but I forsee you will regret the end result :)

Reply:
Something like this ? are you sure ?
```php
            ->addSelect(
                sprintf(
                    'MIN(@%2$s := CASE WHEN @%2$s IS NULL THEN %1$s WHEN %3$s THEN %1$s ELSE @%2$s END) %2$s',
                    $pkColumn,
                    $groupingIdColumn,
                    $groupingCondition
                ),
                sprintf(
                    'CASE WHEN %s THEN CONCAT(\'single-\', %s) ELSE CONCAT(\'range-\', @%s) END %s',
                    $groupingCondition,
                    $pkColumn,
                    $groupingIdColumn,
                    $groupingNameColumn
                )
            )
```